compiler/rustc_index/src/bit_set.rs

```diff
@@ -328,6 +337,409 @@ impl<T: Idx> BitRelations<BitSet<T>> for BitSet<T> {
     }
 }
 
+/// A fixed-size bitset type with a partially dense, partially sparse
+/// representation. The bitset is broken into chunks, and chunks that are all
+/// zeros or all ones are represented and handled very efficiently.
+///
+/// This type is especially efficient for sets that typically have a large
+/// `domain_size` with significant stretches of all zeros or all ones.
+///
+/// `T` is an index type, typically a newtyped `usize` wrapper, but it can also
+/// just be `usize`.
+///
+/// All operations that involve an element will panic if the element is equal
+/// to or greater than the domain size. All operations that involve two bitsets
+/// will panic if the bitsets have differing domain sizes.
+#[derive(Debug, PartialEq, Eq)] // njn: remove Debug?
+pub struct ChunkedBitSet<T> {
+    domain_size: usize,
+
+    /// The chunks. Each one contains exactly CHUNK_BITS values, except the
+    /// last one which contains 1..=CHUNK_BITS values.
+    chunks: Box<[Chunk]>,
+
+    marker: PhantomData<T>,
+}
+
+// njn: comment about chunk_domain_size stored in each variant to keep type
+// size smaller
+#[derive(Clone, Debug, PartialEq, Eq)] // njn: remove Debug?
+enum Chunk {
+    /// A chunk that is all zeros; we don't represent the zeros explicitly.
+    Zeros(u16),
+
+    /// A chunk that is all ones; we don't represent the zeros explicitly.
+    Ones(u16),
+
+    /// A chunk that has a mix of zeros and ones, which are represented
+    /// explicitly and densely. It never has all zeros or all ones.
+    ///
+    /// The length of the word slice is as follows.
+    /// - For all chunks other than the final chunk, the length is CHUNK_WORDS.
+    /// - For the final chunk the length is
+    ///   `num_words(self.chunk_domain_size)`, i.e. the exact number of words
+    ///   needed to cover the domain size.
+    ///
+    /// The second field is the number of 1s set in the chunk.
+    Mixed(u16, Rc<[Word]>, u16), // njn: named fields?
```

Comment:
Comment on why an Rc here? At first glance, probably not obvious.

Also, why `Rc<[Word]>` -- shouldn't we know the chunk width upfront, and this can be `Rc<[Word; CHUNK_WORDS]>`? That'll save space since we skip storing the length.

Reply:
For the second question: we know the chunk width for all chunks except the final one, which can be any length in the range `1..=CHUNK_WORDS`. And lots of sets are small, and so only have one chunk.

Reply:
I tried doing `Rc<Word; CHUNK_WORDS>` and it ends up being better. Good suggestion :)